torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1320,13 +1320,35 @@ struct to_ir {
     pushFrame(comprehension_block);
     WithInsertPoint guard(comprehension_block);
     auto emit_body = [&]() {
-      auto comprehension_out = emitExpr(lc.elt());
+      Value* out = emitExpr(lc.elt());
+
+      // Make sure that any element types are subtypes of the annotated
+      // type
+      std::stringstream ss;
+      if (type_hint &&
+          !out->type()->isSubtypeOfExt(
+              type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << "List type annotation did not match "
```

Comment:
Done! On a side note, I had to allow this behavior in `emitListLiteral` for BC purposes, but I think it really should be an error. How would you feel if I made it an error and simply changed our broken tests? I could stack another PR on top of this as a Better Engineering effort

Reply:
It does seem like a real error. Which test is broken?

Reply:
Yes, I deleted [this](https://github.com/pytorch/pytorch/pull/57137/files#diff-377406d8c2c4a9c34da88772f543dc94a0e81753a0fcb6208cdc579a057e627fL574) test case. It seems to me like what the code tests for is a mistake, but this may be me misunderstanding `torch.jit.interface` (there's no documentation for it at all). Please take a look before I ship this